proposals/306-ipv6-happy-eyeballs.txt

```diff
@@ -0,0 +1,300 @@
+   on most networks in a so-called "dual-stack" configuration. This is to not
+   break existing IPv4-only applications while enabling IPv6 connectivity.
+   However, IPv6 connectivity may be unreliable and clients should be able
+   to connect to the guard using the most reliable technology, whether IPv4
+   or IPv6.
+
+   In ticket #27490, we introduced the option ClientAutoIPv6ORPort which
+   lets a client randomly choose between IPv4 or IPv6. However, this
+   random decision does not take into account unreliable connectivity
+   or falling back to the competing IP version should one be unreliable
+   or unavailable.
+
+   One way to select between IPv4 and IPv6 on a dual-stack network is a
+   so-called "Happy Eyeballs" algorithm as per RFC 8305. In one, a client
+   attempts the preferred IP family, whether IPv4 or IPv6. Should it work,
+   the client sticks with the preferred IP family. Otherwise, the client
+   attempts the alternate version. This means if a dual-stack client has
+   both IPv4 and IPv6, and IPv6 is unreliable, preferred or not, the
+   client uses IPv4, and vice versa. However, if IPv4 and IPv6 are both
+   equally reliable, and IPv6 is preferred, we use IPv6.
+
+   In Proposal 299, we have attempted a IP fallback mechanism using failure
+   counters and preferring IPv4 and IPv6 based on the state of the counters.
+   However, Prop299 was not standard Happy Eyeballs and an alternative,
+   standards-compliant proposal was requested in [P299-TRAC] to avoid issues
+   from complexity caused by randomness.
+
+   This proposal describes a Tor implementation of Happy Eyeballs and is
+   intended as a successor to Proposal 299.
+
+2. Address/Relay Selection
+
+   This section describes the necessary changes for address selection to
+   implement Prop306.
+
+2.1. Extend Info Structure Changes
+
+   To be able to handle Happy Eyeballs in Tor, we will need to modify the
+   data structures used for connections to guards, namely the extend info
+   structure.
+
+   The extend info structure should contain both an IPv4 and an IPv6 address.
+   This will allow us to try IPv4 and the IPv6 addresses should both be
+   available on a relay and the client is dual-stack.
+
+   When parsing relay descriptors and filling in the extend info data
+   structure, we need to fill in both the IPv4 and IPv6 address if they both
+   are available. If only one family is available for a relay (IPv4 or IPv6),
+   we should fill in the address for preferred family and leave the alternate
+   family null.
+
+2.2. Relay Selection Changes
+
+   In Proposal 283, we have allowed microdescriptor consensus documents to
+   contain IPv6 addresses. As clients download microdescriptors, Prop283
+   makes it possible to implement Prop306.
+
+   When we select candidates for the entry guard, we should select at least
+   one relay with an IPv6 address. This makes it possible for an IPv6-only
```

Comment:
One out of 3 primary guards isn't great for redundancy, but we can't afford any more until we have more IPv6 relays. Let's make this a consensus parameter and option like NumPrimaryGuards:
    NumIPv6Guards NUM

    If UseEntryGuards is set to 1, we will try to pick NUM routers that support IPv6 for our primary guard list, which is the set of routers we strongly prefer when connecting to the Tor network. If NUM is 0, we try to learn the number from the guard-n-primary-guards-ipv6 consensus parameter, and default to 1 if the consensus parameter isn’t set. (Default: 0)